Target and Count encodings for categorical features (#3234)
microsoft/LightGBM: docs/Advanced-Topics.rst

Categorical Feature Support
---------------------------

-  LightGBM offers good accuracy with integer-encoded categorical features. LightGBM offers the following approaches to deal with categorical features:

   -  Method 1: Applies `Fisher (1958) <https://www.tandfonline.com/doi/abs/10.1080/01621459.1958.10501479>`__ to find the optimal split over categories as `described here <./Features.rst#optimal-split-for-categorical-features>`__.

   -  Method 2: Encoding categorical features into numerical values. We provide two encoding options:

>       -  **Target encoding**: encode the categorical feature value by the mean of labels of data with the same feature value in the training set. It is easy to overfit the training data if the encoded value of a training data point uses the label of that training data point itself. So we would randomly divide the training data into folds, and when calculating the target encoding for data in one fold, we only consider data in other folds.

Reviewer: ```suggestion
      -  **Target encoding**: encode the categorical feature value by the mean of labels of data with the same feature value in the training set. It is easy to overfit the training data if the encoded value of a training data point uses the label of that training data point itself. So LightGBM randomly divides the training data into folds, and when calculating the target encoding for data in one fold, only considers data in other folds.
```

I think it's important to remove language like "I" or "we" or "you" from documentation talking about LightGBM's behavior, to avoid users being confused and thinking that this doc means they need to do some preprocessing themselves before passing data to LightGBM.

Author: Resolved via fab88db.

-  Categorical features must be encoded as non-negative integers (``int``) less than ``Int32.MaxValue`` (2147483647).
   It is best to use a contiguous range of integers started from zero.

-  Use ``category_encoders`` to specify the methods used to deal with categorical features. We use

   -  ``raw`` to indicate method 1.

   -  ``target[:prior]`` to indicate target encoding in method 2. The ``prior`` is a real number used to smooth the calculation of encoded values. So ``target[:prior]`` is calculated as: ``(sum_label + prior * prior_weight) / (count + prior_weight)``. Here ``sum_label`` is the sum of labels of data in the training set with the same categorical feature value, ``count`` is the total number of data with the same feature value in the training set (the value of count encoding), and ``prior_weight`` is a hyper-parameter. If the prior value is missing, we use the mean of all labels of training data as default prior.

   -  ``count`` to indicate count encoding in method 2.

   Note that the aforementioned methods can be used simultaneously. Different methods are separated by commas.

>    For example ``category_encoders=target:0.5,target:count,raw`` will enable using splits with method 1, and in addition, convert each categorical feature into 3 numerical features. The first one uses target encoding with prior ``0.5``. The second one uses target encoding with default prior, which is the mean of labels of the training data. The third one uses count encoding.

Reviewer 2: I'm a bit confused here. Should we change ``category_encoders=target:0.5,target:count,raw``  to ``category_encoders=target:0.5,target,count,raw``?

Author: Oh, yes. I did not notice that. Thanks!

Author: Resolved via a4ec01f
